### PR TITLE
resPlacement

### PR DESCRIPTION
**Pull request by Esras**
First project PR, hopefully I followed the right standards.

Adds the ksys::res::Placement class. I do think I'm missing two functions, but working on adding them and wanted feedback on where this was right now.

- - -
This change is [Reviewable](https://reviewable.io/reviews/zeldaret/botw/120)

### CODE REVIEW

**Comment by notyourav** [comment]
*[`src/KingSystem/Resource/resPlacement.h` line 17 at r1](https://reviewable.io/reviews/zeldaret/botw/120#-NXNE8PJ2R6rgRRtbjq5:-NXNE8PJ2R6rgRRtbjq6:b-sp7f4m) ([raw file](https://github.com/zeldaret/botw/blob/912821c10aad933d3c122a654c3f6a765b09ed5e/src/KingSystem/Resource/resPlacement.h#L17)):*
> ```objc
>     void registerFactory(sead::Heap* heap);
> 
>     long mVar1;
> ```

I would make these s32

**Review by notyourav — COMMENTED** [review]
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/120)* status: all files reviewed, 1 unresolved discussion (waiting on @Esras)

**Review by leoetlino — CHANGES_REQUESTED** [review]
taking a look at the vtables, it looks like there are a few functions you have reimplemented but not listed in the CSV:

- F42740 is res::Placement::checkDerivedRuntimeTypeInfo
  - F42804 (the function which is called by checkDerivedRuntimeTypeInfo) is res::Placement::checkDerivedRuntimeTypeInfoStatic
- F42748 is res::Placement::getRuntimeTypeInfo
- F427A4 is res::Placement::~Placement (the D0 deleting destructor; refer to https://botw.link/contribute/cpp-primer#virtual-destructors if you're not sure what that means)
- F427C8 is res::Placement::doCreate_ (okay, looks like the two members are definitely pointers)
- F427D4 is a non-virtual thunk to res::Placement::~Placement
- F427DC is the same (but a thunk to the D0 destructor)

you should take a look at the vtable of `res::EntryFactory<res::Placement>` to ensure you list all of its functions in the CSV (you get all of those functions generated for you for free, would be a shame not to make the decompiled code% higher :P)

> Reviewed 4 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/120#-:-NXRSndZDpssuvGZrm4q:bhjtntx)* status: all files reviewed, 4 unresolved discussions (waiting on @Esras and @notyourav)

___
*[`src/KingSystem/Resource/resPlacement.h` line 15 at r1](https://reviewable.io/reviews/zeldaret/botw/120#-NXRSUvJ2Usaabgr4Lj3:-NXRSUvJ2Usaabgr4Lj4:b-z23je8) ([raw file](https://github.com/zeldaret/botw/blob/912821c10aad933d3c122a654c3f6a765b09ed5e/src/KingSystem/Resource/resPlacement.h#L15)):*
> ```objc
>     ~Placement() override;
> 
>     void registerFactory(sead::Heap* heap);
> ```

It'd make more sense for this to be a static member function
___
*[`src/KingSystem/Resource/resPlacement.h` line 17 at r1](https://reviewable.io/reviews/zeldaret/botw/120#-NXNE8PJ2R6rgRRtbjq5:-NXRSIfs1CoB34p7-nl3:bu76hwt) ([raw file](https://github.com/zeldaret/botw/blob/912821c10aad933d3c122a654c3f6a765b09ed5e/src/KingSystem/Resource/resPlacement.h#L17)):*
<details><summary><i>Previously, notyourav wrote…</i></summary><blockquote>

I would make these s32
</blockquote></details>

long is 64-bit, not 32-bit. I think these are probably pointers - not sure what type though, so I'd just make them void* (e.g. `void* mUnk = nullptr;`)
___
*[`src/KingSystem/Resource/resPlacement.cpp` line 16 at r1](https://reviewable.io/reviews/zeldaret/botw/120#-NXRSbj54QAOL6_uaFVI:-NXRSbj54QAOL6_uaFVJ:b-o09o3t) ([raw file](https://github.com/zeldaret/botw/blob/912821c10aad933d3c122a654c3f6a765b09ed5e/src/KingSystem/Resource/resPlacement.cpp#L16)):*
> ```cpp
> 
>     res::registerEntryFactory(new (heap) EntryFactory<Placement>(1.0, sizeof(Placement)),
>                               "bPlacement");
> ```

bplacement?
___
*[`src/KingSystem/Resource/resPlacement.cpp` line 19 at r1](https://reviewable.io/reviews/zeldaret/botw/120#-NXRSP3K4Q-zqvuaXpSl:-NXRSP3LCGKzgejw9t8a:b-9svq62) ([raw file](https://github.com/zeldaret/botw/blob/912821c10aad933d3c122a654c3f6a765b09ed5e/src/KingSystem/Resource/resPlacement.cpp#L19)):*
> ```cpp
> }
> 
> Placement::Placement() {
> ```

this can just be `Placement::Placement() = default;` after the above comment is addressed

**Comment by Esras** [comment]
Thanks for the detailed feedback!

I think I got thrown off because there's the VTable for the Placement struct, the EntryFactory<Placement> struct, and then one more for just the Placement destructors, it seems? I might be missing one more because I'm not sure where the `ksys::res::EntryFactory<ksys::res::Placement>::checkDerivedRuntimeTypeInfo` function is and I can't seem to find where the registerFactory function would live (I'm not sure it does, the registration seems to live in `res::ResourceMgrTask::init()` (F38D5C).

**Review by leoetlino — COMMENTED** [review]
> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/120)* status: 2 of 4 files reviewed, all discussions resolved (waiting on @Esras)

**Comment by leoetlino** [comment]
* 0x0000007100f427c8 should have the mangled name of ksys::res::Placement::doCreate_ listed
* 0x0000007100f427d4 is not doDestroy, it's a non-virtual thunk to the D1/D2 destructor (this exists because ksys::res::Resource uses multiple inheritance)
* 0x0000007100f427dc is a non-virtual thunk to the D0 destructor

You'll find the correct symbols to use for the non-virtual thunks by running `tools/listsym res::Placement`

>  I might be missing one more because I'm not sure where the `ksys::res::EntryFactory<ksys::res::Placement>::checkDerivedRuntimeTypeInfo` function is

I think that's 0x0000007100f4259c. Sometimes checkDerivedRuntimeTypeInfoStatic gets inlined, which means you'll only see checkDerivedRuntimeTypeInfo.

> and I can't seem to find where the registerFactory function would live (I'm not sure it does, the registration seems to live in res::ResourceMgrTask::init() (F38D5C).

Yeah, I'm not sure registerFactory actually exists. As you've noticed, the registration is done in res::ResourceMgrTask. I think it's fine to keep registerFactory for the time being to force the compiler to emit the res::Placement and EntryFactory functions.

File: data/uking_functions.csv [hunk 1]
```diff
@@ -81914,20 +81914,20 @@ Address,Quality,Size,Name
...
-0x0000007100f42740,U,000008,
-0x0000007100f42748,U,000092,
-0x0000007100f427a4,U,000036,
-0x0000007100f427c8,U,000012,
-0x0000007100f427d4,U,000008,
-0x0000007100f427dc,U,000040,
-0x0000007100f42804,U,000372,
+0x0000007100f4250c,O,000068,_ZN4ksys3res12EntryFactoryINS0_9PlacementEED2Ev
+0x0000007100f42550,O,000076,_ZN4ksys3res12EntryFactoryINS0_9PlacementEED0Ev
+0x0000007100f4259c,W,000132,_ZN4ksys3res12EntryFactoryINS0_9PlacementEE33checkDerivedRuntimeTypeInfoStaticEPKN4sead15RuntimeTypeInfo9InterfaceE
```
Comment (leoetlino) [inline data/uking_functions.csv:81919]: checkDerivedRuntimeTypeInfo, notcheckDerivedRuntimeTypeInfoStatic

**Review by leoetlino — APPROVED** [review]

**Review by leoetlino — APPROVED** [review]
> Reviewed 2 of 2 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/120)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Esras)

**Pull request merged**